keps/sig-testing/20190118-breaking-apart-the-kubernetes-test-tarball.md

```diff
@@ -0,0 +1,221 @@
+- It slightly complicates downloading of test dependencies
+2. Duplicate these sources into each binary-specific tarball.
+- Simplifies test dependency distribution
+- Portable sources are small (as of v1.13.2, approximately 2.7MB uncompressed
+  or about 186KB compressed) so duplication isn't a huge worry
+- Complicates extraction of tarballs with existing scripts, since they assume
+  everything is platform-specific
+
+It's not immediately obvious which of these is the better approach based on the
+tradeoffs.
+
+### Updating dependencies on `kubernetes-test.tar.gz`
+
+Currently the CI workflows and `kubetest` use the `cluster/get-kube.sh` and
+`cluster/get-kube-binaries.sh` scripts to download all artifacts, and
+conveniently `get-kube-binaries.sh` is versioned with the release artifacts in
+`kubernetes.tar.gz`, so simply making `get-kube-binaries.sh` aware of the new
+tarballs should be sufficient for most CI and developer needs.
+
+Because the test tarball includes binaries used both on the host running tests
+(such as the `e2e` binary), as well as binaries which may run the nodes
+(`e2e.node`), we would need to make sure to download binary test artifacts
+targeting the host platform, node platform, and possibly server platform.
+
+A quick search reveals a few other uses of `kubernetes-test.tar.gz`, mostly in
+`cluster/`. We can update these to use the platform-specific tarballs, possibly
+with a fallback to the mondo-tarball if worried about versioning.
+
+#### Dependencies outside the Kubernetes organization
+
+Searching GitHub for references to `kubernetes-test.tar.gz` largely returns
+forks of the main kubernetes repository (including some very old forks,
+identifiable by the script `e2e-from-release.sh`). Since these forks are not
+likely to depend on upstream release artifacts, we can ignore them.
+
+The Samsung SDS CNCT kraken-lib repository has a reference to `kubernetes-test.tar.gz`
+in its [conformance test script](https://github.com/samsung-cnct/kraken-lib/blob/aceab16c316bafcdb1f542dc67876dd2e5279f6b/build-scripts/conformance-tests#L16),
+but this repo is also marked deprecated and read-only, and there have been no
+changes since July 2018.
+
+In vmware/simple-k8s-test-env, the `sk8.sh` file uses
+[`kubernetes-test.tar.gz`](https://github.com/vmware/simple-k8s-test-env/blob/master/sk8.sh#L4481),
+and this repo seems actively maintained, so we should make sure this continues
+working.
+
+The [reference](https://github.com/knative/test-infra/blob/8ef3dc1c2ed07e64024bc68c9dbd1a2e10e9e975/scripts/e2e-tests.sh#L118-L120)
+to `kubernetes-test.tar.gz` in knative/test-infra is hilarious.
+
+### Risks and Mitigations
+
+It's hard to tell who uses these test tarballs outside the core project and
+without tools like `kubetest`. We'll need to broadcast this change widely
+so that any downstream users are aware of the incompatible changes.
+
+## Graduation Criteria
+
+All goals are met?
```

Comment:
Based on the risk of alienating downstream, it's unclear to me whether we're trying to do a clean break this release, or publish both split and mondo

From a "is this a checklist the release team can consume" perspective, I would feel comfortable if I saw some testgrid dashboards that proved the split tarballs are being consumed (by different arches). 

Something like kubetest-driven job that definitely downloads a test tarball instead of building the tests itself, and preferably more than one arch.

Haven't looked to see how much of these we can change:
- https://testgrid.k8s.io/conformance-arm#Arm64%20AWS
- https://testgrid.k8s.io/sig-node-arm64#conformance
- https://testgrid.k8s.io/sig-cluster-lifecycle-multi-platform#Summary (though everything here is stale)

Reply:
I think most of our CI e2e jobs are kubetest-driven and download all artifacts rather than building them at test-time; it's only the PR jobs that build first, and even those use the `cluster/get-kube.sh` scripts to separate out the phases.

I've updated this section with some more language.